Content.Shared/_Impstation/Replicator/SharedReplicatorNestSystem.cs

```diff
@@ -0,0 +1,336 @@
+                continue;
+
+            if (!TryComp<MindContainerComponent>(replicator, out var mindContainer) || mindContainer.Mind == null)
+                continue;
+
+            comp.TargetUpgradeStage++;
+
+            var targetAction = comp.TargetUpgradeStage == 1 ? comp.Level2Action : comp.Level3Action;
+
+            if (!mindContainer.HasMind)
+                _actions.AddAction(replicator, targetAction);
+            else if (mindContainer.Mind != null)
+                _actionContainer.AddAction((EntityUid)mindContainer.Mind, targetAction);
+        }
+    }
+
+    public void OnUpgrade2(Entity<ReplicatorComponent> ent, ref ReplicatorUpgrade2ActionEvent args)
+    {
+        // don't run this clientside
+        if (_net.IsClient || !_timing.IsFirstTimePredicted)
+            return;
+
+        UpgradeReplicator(ent, 2);
+
+        QueueDel(ent);
+        QueueDel(args.Action);
+    }
+
+    public void OnUpgrade3(Entity<ReplicatorComponent> ent, ref ReplicatorUpgrade3ActionEvent args)
+    {
+        // don't run this clientside
+        if (_net.IsClient || !_timing.IsFirstTimePredicted)
+            return;
+
+        UpgradeReplicator(ent, 3);
+
+        QueueDel(ent);
+        QueueDel(args.Action);
+    }
+
+    public void UpgradeReplicator(Entity<ReplicatorComponent> ent, int desiredLevel)
+    {
+        var xform = Transform(ent);
+
+        // if adding more stages, maybe make this a switch.
+        var nextStage = desiredLevel == 2 ? ent.Comp.Level2Id : ent.Comp.Level3Id;
+
+        var upgraded = Spawn(nextStage, xform.Coordinates);
+        var upgradedComp = EnsureComp<ReplicatorComponent>(upgraded);
+        upgradedComp.RelatedReplicators = ent.Comp.RelatedReplicators;
+        upgradedComp.TargetUpgradeStage = ent.Comp.TargetUpgradeStage;
+
+        if (ent.Comp.MyNest != null)
+        {
+            var nestComp = EnsureComp<ReplicatorNestComponent>((EntityUid)ent.Comp.MyNest);
+            nestComp.SpawnedMinions.Remove(ent);
+            nestComp.SpawnedMinions.Add(upgraded);
+        }
+
+        if (!_mind.TryGetMind(ent, out var mind, out _))
+            return;
+
+        _mind.TransferTo(mind, upgraded);
+
+        return;
+    }
+
+    private void Embiggen(Entity<ReplicatorNestComponent> ent)
+    {
+        var ev = new ReplicatorNestEmbiggenedEvent(ent);
+        RaiseLocalEvent(ent, ref ev);
+    }
+}
+
+public sealed partial class ReplicatorSpawnNestActionEvent : InstantActionEvent
+{
+
+}
+
+public sealed partial class ReplicatorUpgrade2ActionEvent : InstantActionEvent
+{
+
+}
+
+public sealed partial class ReplicatorUpgrade3ActionEvent : InstantActionEvent
+{
+
+}
+
+[ByRefEvent]
+public sealed partial class ReplicatorNestEmbiggenedEvent : EntityEventArgs
+{
+    public Entity<ReplicatorNestComponent> Ent { get; set; }
+    public ReplicatorNestEmbiggenedEvent(Entity<ReplicatorNestComponent> ent)
```

Comment:
could save a few lines by changing line 329 to read:
`public sealed partial class ReplicatorNestEmbiggenedEvent(Entity<ReplicatorNestComponent> ent) : EntityEventArgs`
then this bit can get removed